### PR TITLE
[Leveling] low,high with tests

### PR DESCRIPTION
**Pull request by ThomasRetornaz**
Rework leveling PR for integrating strel

### CODE REVIEW

File: src/leveling.jl [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        end
+    end
+    #Post-processing, so that we have nicer levelings
+    #(L-= inf( ref, dilate(mark)) , and L+=sup(ref,erode(mark)),
+    #and thus L(ref,mark)=L-( L+(ref,mark),mark) and they commute)
+    for i in R
+        if out[i] > ref[i]
+            out[i] = ref[i]
+        end
+    end
```
Comment (johnnychen94) [inline src/leveling.jl:93]:
```suggestion
    out .= max.(out, ref)
```

File: src/leveling.jl [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    #Post-processing, so that we have nicer levelings
+    #(L-= inf( ref, dilate(mark)) , and L+=sup(ref,erode(mark)),
+    #and thus L(ref,mark)=L-( L+(ref,mark),mark) and they commute)
+    for i in R
+        if out[i] < ref[i]
+            out[i] = ref[i]
+        end
+    end
```
Comment (johnnychen94) [inline src/leveling.jl:185]:
```suggestion
    out .= max.(out, ref)
```

File: src/leveling.jl [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    # declare priority queue, store position of pixel associated with priority 
+    pq = PriorityQueue{CartesianIndex{N},T}()
+    R = CartesianIndices(axes(out))
+    for i in R
+        @inbounds if out[i] < ref[i]
+            @inbounds current_max = out[i]
+            for Δi in se # examine neighborhoods
+                ii = i + Δi
+                if checkbounds(Bool, R, ii) #check that we are in the image
+                    @inbounds current_max = max(current_max, out[ii])
```
Comment (johnnychen94) [inline src/leveling.jl:60]: IIUC, when you add an `@inbounds` at the outer place, then it applies to all the inner regions. Thus you don't need to add `@inbounds` for every place.

This holds true as long as it is the same function. That said, if there's another function that you want to pass `@inbounds`, you can use `Base.@propagate_inbounds` to define that _inline_-ed function:

```julia
@inline f(x, i) = x[i]
Base.@propagate_inbounds g(x, i) = x[i]

hf() = @inbounds f([1], 2) # this errors
hf()

hg() = @inbounds g([1], 2) # this doesn't
hg()
```

The `@inbounds` macro only applies when it's inside a function, which is why we need `hg()`/`hf()` in above example.

**Review by johnnychen94 — APPROVED** [review]
I didn't check the correctness, but the overall codes look great to me!

@ThomasRetornaz I'm currently working full-time in [TongYuan](https://github.com/Suzhou-Tongyuan) so would be kept super busy. I've sent you an invitation so feel free to merge when you're confident.

**Comment by ThomasRetornaz** [comment]
> I didn't check the correctness, but the overall codes look great to me!
> 
> @ThomasRetornaz I'm currently working full-time in [TongYuan](https://github.com/Suzhou-Tongyuan) so would be kept super busy. I've sent you an invitation so feel free to merge when you're confident.

Hi glad to know that you found i hope a interesting position :)
I will check your remarks
King regards
TR

**Pull request merged**